Terraform crashes after removing a revision from rke_cluster_template

I accidentally deleted an RKE cluster template revision on the UI. Then I reran Terraform in order to recreate it. However, Terraform crashed with the error:
```
Error: rpc error: code = Unavailable desc = transport is closing
panic: runtime error: index out of range [4] with length 4
```

IMO, it should be possible to re-setup the infrastructure after it drifts way from the Terraform code.

@rawmind0 : Sure, sorry I thought it was easier for your to see only the 2 important lines. This is the full log:
```
Error: rpc error: code = Unavailable desc = transport is closing
panic: runtime error: index out of range [4] with length 4
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: goroutine 16 [running]:
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: github.com/terraform-providers/terraform-provider-rancher2/rancher2.flattenClusterTemplateRevisions(0xc0007ca000, 0x4, 0x4, 0xc000a76e40, 0x1c, 0xc000022660, 0x6, 0x6, 0x1a7cdc0, 0xc000734700, ...)
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 	/go/src/github.com/rancher/terraform-provider-rancher2/rancher2/structure_cluster_template.go:131 +0x1186
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: github.com/terraform-providers/terraform-provider-rancher2/rancher2.flattenClusterTemplate(0xc00013ca80, 0xc000734700, 0xc0007ca000, 0x4, 0x4, 0x4, 0x0)
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 	/go/src/github.com/rancher/terraform-provider-rancher2/rancher2/structure_cluster_template.go:215 +0x127
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: github.com/terraform-providers/terraform-provider-rancher2/rancher2.resourceRancher2ClusterTemplateRead(0xc00013ca80, 0x1e91200, 0xc0004d8240, 0xc00013ca80, 0x0)
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 	/go/src/github.com/rancher/terraform-provider-rancher2/rancher2/resource_rancher2_cluster_template.go:208 +0x2e6
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc000150ab0, 0xc0007623c0, 0x1e91200, 0xc0004d8240, 0xc000121128, 0x0, 0x0)
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 	/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.14.0/helper/schema/resource.go:460 +0x119
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ReadResource(0xc000120d40, 0x22cb100, 0xc000d7a030, 0xc0001ee180, 0xc000120d40, 0xc000d7a030, 0xc000355b78)
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 	/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.14.0/internal/helper/plugin/grpc_provider.go:525 +0x3d8
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ReadResource_Handler(0x1dd8bc0, 0xc000120d40, 0x22cb100, 0xc000d7a030, 0xc0001ee120, 0x0, 0x22cb100, 0xc000d7a030, 0xc000e14000, 0x50ee)
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 	/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.14.0/internal/tfplugin5/tfplugin5.pb.go:3269 +0x217
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: google.golang.org/grpc.(*Server).processUnaryRPC(0xc000702a80, 0x22e4360, 0xc000602600, 0xc0002b3c00, 0xc0006dc300, 0x32b6550, 0x0, 0x0, 0x0)
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 	/go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:1024 +0x501
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: google.golang.org/grpc.(*Server).handleStream(0xc000702a80, 0x22e4360, 0xc000602600, 0xc0002b3c00, 0x0)
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 	/go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:1313 +0xd3f
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00004a170, 0xc000702a80, 0x22e4360, 0xc000602600, 0xc0002b3c00)
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 	/go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:722 +0xa1
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: created by google.golang.org/grpc.(*Server).serveStreams.func1
2020-11-11T14:23:07.206Z [DEBUG] plugin.terraform-provider-rancher2_v1.10.4: 	/go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:720 +0xa1
2020-11-11T14:23:07.208Z [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2020/11/11 14:23:07 [ERROR] module.rancher: eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2020/11/11 14:23:07 [ERROR] module.rancher: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2020/11/11 14:23:07 [TRACE] [walkRefresh] Exiting eval tree: module.rancher.rancher2_cluster_template.default
2020/11/11 14:23:07 [TRACE] vertex "module.rancher.rancher2_cluster_template.default": visit complete
2020-11-11T14:23:07.208Z [DEBUG] plugin: plugin process exited: path=/builds/it-ops/rancher-setup/environments/ops/.terraform/plugins/linux_amd64/terraform-provider-rancher2_v1.10.4 pid=633 error="exit status 2"
2020/11/11 14:23:07 [TRACE] vertex "module.rancher.rancher2_cluster_template.default": dynamic subgraph encountered errors
2020/11/11 14:23:07 [TRACE] vertex "module.rancher.rancher2_cluster_template.default": visit complete
2020/11/11 14:23:07 [TRACE] dag/walk: upstream of "module.rancher.provider.rancher2 (close)" errored, so skipping
2020/11/11 14:23:07 [TRACE] dag/walk: upstream of "module.rancher.output.cluster_template_id" errored, so skipping
2020/11/11 14:23:07 [TRACE] dag/walk: upstream of "output.cluster_template_id" errored, so skipping
2020/11/11 14:23:07 [TRACE] dag/walk: upstream of "module.rancher.output.cluster_template_revision_id" errored, so skipping
2020/11/11 14:23:07 [TRACE] dag/walk: upstream of "output.cluster_template_revision_id" errored, so skipping
2020/11/11 14:23:07 [TRACE] dag/walk: upstream of "root" errored, so skipping
2020-11-11T14:23:07.209Z [DEBUG] plugin: plugin exited
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.
When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.
SECURITY WARNING: the "crash.log" file that was created may contain 
sensitive information that must be redacted before it is safe to share 
on the issue tracker.
[1]: https://github.com/hashicorp/terraform/issues
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,17 @@
+## 1.10.6 (November 11, 2020)
+
+FEATURES:
+
+
+
+ENHANCEMENTS:
+
+
+
+BUG FIXES:
+
+* Fixed `flattenClusterTemplateRevisions` func to avoid crash on `rancher2_cluster_template` resource at some circunstances
+
 ## 1.10.5 (November 11, 2020)
 
 FEATURES:

--- a/rancher2/structure_cluster_template.go
+++ b/rancher2/structure_cluster_template.go
@@ -123,16 +123,24 @@ func flattenClusterTemplateRevisions(input []managementClient.ClusterTemplateRev
 			}
 		}
 	}
-	sortedInput := make([]managementClient.ClusterTemplateRevision, len(input))
+	inputLen := len(input)
+	pLen := len(p)
+	sortedInput := make([]managementClient.ClusterTemplateRevision, inputLen)
 	newCTR := []managementClient.ClusterTemplateRevision{}
 	lastIndex := 0
 	for i := range sortedInput {
 		if v, ok := pIndexID[input[i].ID]; ok {
+			if v > i && pLen > inputLen {
+				v = v - (v - i)
+			}
 			sortedInput[v] = input[i]
 			lastIndex++
 			continue
 		}
 		if v, ok := pIndexName[input[i].Name]; ok {
+			if v > i && pLen > inputLen {
+				v = v - (v - i)
+			}
 			sortedInput[v] = input[i]
 			lastIndex++
 			continue

